### PR TITLE
perf(core): optimize queryVerticesByIds for only-one-id query

### PR DESCRIPTION
**Pull request by javeme**
Change-Id: Ic6e01eb8e2b4c97c5263cfa0f78c58b6033435da

### CODE REVIEW

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java [hunk 1]
```diff
@@ -775,40 +784,103 @@ protected Iterator<Vertex> queryVerticesByIds(Object[] vertexIds, boolean adjace
...
+                if (this.removedVertices.containsKey(id)) {
+                    // The record has been deleted
+                    tryQueryBackend = false;
+                } else if ((vertex = this.addedVertices.get(id)) != null ||
+                           (vertex = this.updatedVertices.get(id)) != null) {
+                    // Found from local tx
+                    tryQueryBackend = false;
+                    if (vertex.expired()) {
+                        vertex = null;
+                    } else {
```
Comment (imbajin) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java:818]: Consider extracting the local transaction lookup logic to reduce code duplication between single-ID and multi-ID paths. This pattern repeats in both vertex and edge queries:

```java
private HugeVertex findVertexInLocalTx(Id id) {
    if (this.removedVertices.containsKey(id)) {
        return null; // Removed
    }
    HugeVertex vertex = this.addedVertices.get(id);
    if (vertex == null) {
        vertex = this.updatedVertices.get(id);
    }
    return (vertex != null && !vertex.expired()) ? vertex : null;
}
```

This would simplify the control flow and make the code more maintainable.

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java [hunk 2]
```diff
@@ -775,40 +784,103 @@ protected Iterator<Vertex> queryVerticesByIds(Object[] vertexIds, boolean adjace
...
-            if (id == null || this.removedVertices.containsKey(id)) {
-                // The record has been deleted
-                continue;
-            } else if ((vertex = this.addedVertices.get(id)) != null ||
-                       (vertex = this.updatedVertices.get(id)) != null) {
-                if (vertex.expired()) {
-                    continue;
+        if (vertexIds.length == 1) {
+            Id id = HugeVertex.getIdValue(vertexIds[0]);
+
```
Comment (imbajin) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java:798]: The multiple boolean flags (`tryQueryBackend`, `foundLocal`) make the control flow complex. Consider refactoring into a more readable structure:

```java
private QueryResult<HugeVertex> querySingleVertex(Id id, HugeType type) {
    if (id == null) {
        return QueryResult.empty();
    }
    
    HugeVertex localVertex = findVertexInLocalTx(id);
    if (localVertex != null) {
        return QueryResult.fromLocal(localVertex);
    }
    
    if (isRemovedFromTx(id)) {
        return QueryResult.empty();
    }
    
    return QueryResult.fromBackend(queryBackendVertex(id, type));
}
```

This would eliminate the complex flag management and make the logic clearer.

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java [hunk 3]
```diff
@@ -775,40 +784,103 @@ protected Iterator<Vertex> queryVerticesByIds(Object[] vertexIds, boolean adjace
         return this.queryVerticesByIds(vertexIds, adjacentVertex, checkMustExist, HugeType.VERTEX);
     }
 
+    @Watched(prefix = "graph")
```
Comment (imbajin) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java:787]: Missing test coverage for the single-ID optimization. Consider adding tests for:

1. **Performance verification**: Ensure single-ID queries are faster than multi-ID queries
2. **Correctness tests**: 
   - Single ID found in local transaction (added/updated)
   - Single ID removed from local transaction  
   - Single ID not found anywhere
   - Single null ID handling
3. **Edge cases**: Single ID with expired vertex

Example test structure:
```java
@Test
public void testSingleVertexQueryOptimization() {
    // Test single vertex query performance and correctness
    Id vertexId = IdGenerator.of("test-vertex");
    
    // Test local transaction lookup
    HugeVertex localVertex = constructVertex(vertexId);
    tx.addVertex(localVertex);
    
    Iterator<Vertex> result = tx.queryVertices(vertexId);
    assertThat(result).hasNext();
    assertThat(result.next()).isEqualTo(localVertex);
}
```

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java [hunk 4]
```diff
@@ -775,40 +784,103 @@ protected Iterator<Vertex> queryVerticesByIds(Object[] vertexIds, boolean adjace
...
     protected Iterator<Vertex> queryVerticesByIds(Object[] vertexIds, boolean adjacentVertex,
                                                   boolean checkMustExist, HugeType type) {
         Query.checkForceCapacity(vertexIds.length);
 
-        // NOTE: allowed duplicated vertices if query by duplicated ids
-        List<Id> ids = InsertionOrderUtil.newList();
-        Map<Id, HugeVertex> vertices = new HashMap<>(vertexIds.length);
+        List<Id> ids;
+        Map<Id, HugeVertex> vertices;
+        boolean verticesUpdated = this.verticesInTxSize() > 0;
```
Comment (imbajin) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java:794]: Consider optimizing the `verticesUpdated` check. Currently it calls `this.verticesInTxSize() > 0` which may iterate through all transaction maps. For single-ID queries, you could check more efficiently:

```java
// More efficient check for single ID
boolean hasLocalChanges = this.removedVertices.containsKey(id) || 
                         this.addedVertices.containsKey(id) || 
                         this.updatedVertices.containsKey(id);
```

This would be O(1) instead of potentially O(n) for the transaction size calculation.

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java [hunk 5]
```diff
@@ -934,57 +1012,121 @@ public Edge queryEdge(Object edgeId) {
...
+            HugeEdge edge = null;
+            if (id != null && edgesUpdated) {
+                if (this.removedEdges.containsKey(id)) {
+                    // The record has been deleted
+                    tryQueryBackend = false;
+                } else if ((edge = this.addedEdges.get(id)) != null ||
+                           (edge = this.updatedEdges.get(id)) != null) {
+                    // Found from local tx
+                    tryQueryBackend = false;
+                    if (edge.expired()) {
```
Comment (imbajin) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java:1047]: The same optimization pattern is duplicated for edges and vertices. Consider creating a generic helper method to reduce code duplication:

```java
private <T extends HugeElement> Map<Id, T> querySingleElement(
    Id id, 
    HugeType type,
    Map<Id, T> added,
    Map<Id, T> updated, 
    Map<Id, T> removed,
    Function<IdQuery, Iterator<T>> backendQuery) {
    
    if (id == null) {
        return ImmutableMap.of();
    }
    
    // Check local transaction state
    if (removed.containsKey(id)) {
        return ImmutableMap.of();
    }
    
    T element = added.get(id);
    if (element == null) {
        element = updated.get(id);
    }
    
    if (element != null && !element.expired()) {
        return ImmutableMap.of(id, element);
    }
    
    // Query backend
    IdQuery query = new IdQuery.OneIdQuery(type, id);
    T backendElement = QueryResults.one(backendQuery.apply(query));
    return backendElement == null ? ImmutableMap.of() : ImmutableMap.of(id, backendElement);
}
```

This would eliminate the duplication between vertex and edge query methods.
